lib/executor/variable_arrival_rate_test.go

```diff
@@ -0,0 +1,363 @@
+			},
+			{
+				Duration: types.NullDurationFrom(0),
+				Target:   null.IntFrom(100),
+			},
+			{
+				Duration: types.NullDurationFrom(time.Minute),
+				Target:   null.IntFrom(100),
+			},
+		},
+	}
+	var es *lib.ExecutionSegment
+	changes := config.getPlannedRateChanges(es)
+	require.Equal(t, len(changes), cap(changes))
+	require.Equal(t, 2, len(changes))
+	require.Equal(t, time.Duration(0), changes[0].timeOffset)
+	require.Equal(t, types.NullDurationFrom(time.Millisecond*20), changes[0].tickerPeriod)
+
+	require.Equal(t, time.Minute, changes[1].timeOffset)
+	require.Equal(t, types.NullDurationFrom(time.Millisecond*10), changes[1].tickerPeriod)
+}
+
+// helper function to calculate the expected rate change at a given time
+func calculateTickerPeriod(current, start, duration time.Duration, from, to int64) types.Duration {
+	var coef = big.NewRat(
+		(current - start).Nanoseconds(),
+		duration.Nanoseconds(),
+	)
+
+	var oneRat = new(big.Rat).Mul(big.NewRat(from-to, 1), coef)
+	oneRat = new(big.Rat).Sub(big.NewRat(from, 1), oneRat)
+	oneRat = new(big.Rat).Mul(big.NewRat(int64(time.Second), 1), new(big.Rat).Inv(oneRat))
+	return types.Duration(new(big.Int).Div(oneRat.Num(), oneRat.Denom()).Int64())
+}
+
+func TestGetPlannedRateChangesZeroDurationStart(t *testing.T) {
+	// TODO: Make multiple of those tests
+	t.Parallel()
+	var config = VariableArrivalRateConfig{
+		TimeUnit:  types.NullDurationFrom(time.Second),
+		StartRate: null.IntFrom(0),
+		Stages: []Stage{
+			{
+				Duration: types.NullDurationFrom(0),
+				Target:   null.IntFrom(50),
+			},
+			{
+				Duration: types.NullDurationFrom(time.Minute),
+				Target:   null.IntFrom(50),
+			},
+			{
+				Duration: types.NullDurationFrom(0),
+				Target:   null.IntFrom(100),
+			},
+			{
+				Duration: types.NullDurationFrom(time.Minute),
+				Target:   null.IntFrom(100),
+			},
+			{
+				Duration: types.NullDurationFrom(time.Minute),
+				Target:   null.IntFrom(0),
+			},
+		},
+	}
+
+	var es *lib.ExecutionSegment
+	changes := config.getPlannedRateChanges(es)
+	require.Equal(t, len(changes), cap(changes))
+	var expectedTickerPeriod types.Duration
+	for i, change := range changes {
+		switch {
+		case change.timeOffset == 0:
+			expectedTickerPeriod = types.Duration(20 * time.Millisecond)
+		case change.timeOffset == time.Minute*1:
+			expectedTickerPeriod = types.Duration(10 * time.Millisecond)
+		case change.timeOffset < time.Minute*3:
+			expectedTickerPeriod = calculateTickerPeriod(change.timeOffset, 2*time.Minute, time.Minute, 100, 0)
+		case change.timeOffset == time.Minute*3:
+			expectedTickerPeriod = 0
+		default:
+			t.Fatalf("this shouldn't happen %d index %+v", i, change)
+		}
+		require.Equal(t, time.Duration(0),
+			change.timeOffset%minIntervalBetweenRateAdjustments, "%d index %+v", i, change)
+		require.Equal(t, change.tickerPeriod.Duration, expectedTickerPeriod, "%d index %+v", i, change)
+	}
+}
+
+func TestGetPlannedRateChanges2(t *testing.T) {
```

Comment:
Could we rename this to something other than "...Changes2"? :) Or could it be a sub-test alongside `TestGetPlannedRateChangesZeroDurationStart` (so have a single `TestGetPlannedRateChanges` parent test)?

Reply:
:man_facepalming: the `2` was added as I wanted to run only this one (without the other ones) and it was faster to change how this was named ... so I am more for renaming this back to `TestGetPlannedRateChanges`. 

I am not a big fan of HUGE tests with a lot of nested subTests as it's usually just gets a lot of unneeded indentation and ... even less readable ... (take a look at the test in [modules/http](https://github.com/loadimpact/k6/blob/master/js/modules/k6/http/request_test.go)). 
Also this (and a lot of the other tests) should become a [table drive tests](https://github.com/golang/go/wiki/TableDrivenTests) at some point ... probably when we start seeing bugs and writing tests for them ;)